src/content/learning/using-durable-objects.md

```diff
@@ -0,0 +1,216 @@
+
+The first variable passed to the class constructor contains state specific to the Durable Object, including methods for accessing storage.
+
+```js
+
+class DurableObjectExample {
+    constructor(state, environment){
+
+    }
+}
+
+```
+
+Workers communicate with a Durable Object via the fetch API.  Like any other Worker, a Durable Object listens for incoming Fetch events by registering an event handler.
+
+```js
+
+class DurableObjectExample {
+    constructor(state, environment){
+    }
+    
+    addEventListener("fetch", (event) => {
+        event.respondWith(handle(event.request));
+    });
+ 
+    async function handle(request) {
+        return new Response('Hello World');
+    }
+
+}
+
+```
+
+A Worker can pass information to a Durable Object via headers, the HTTP method, the Request body, or the Request URI.
+
+### Accessing Storage from a Durable Object
+
+Durable Objects gain access to a TODO: fill in[persistent storage API]() via the first parameter passed to the Durable Object constructor.  While access to a Durable Object is single-threaded, it's important to remember that there can still be race conditions across multiple requests.
+
+```js
+
+class DurableObjectExample {
+    constructor(state, environment){
+        this.state = state
+    }
+    
+    addEventListener("fetch", (event) => {
+        event.respondWith(handle(event.request));
+    });
+ 
+    async function handle(request) {
+        let ip = request.headers.get('CF-Connecting-IP');
+        let data = request.text();
+        let storagePromise = this.state.storage.set(ip, data);
+        await storagePromise;
+        return new Response(ip + ' stored ' + data);
+    }
+
+}
+
+```
+
+Single statement transactions are always transactional.  More complex use cases can wrap multiple storage statements in a transaction.  
+
+```js
+
+class DurableObjectExample {
+    constructor(state, environment){
+        this.state = state
+    }
+    
+    addEventListener("fetch", (event) => {
+        event.respondWith(handle(event.request));
+    });
+ 
+    async function handle(request) {
+        // TODO: txn example
+    }
+
+}
+
+```
+
+Transactions can fail if they conflict with a concurrent transaction from the same Durable Object.  Transactions are transparently and automatically retried once before returning a failure.
+
+
+## Binding the class to a Worker
+
+As with Workers KV, the class namespace must be created and then bound into a Worker before it can be used.
+
+First, upload the script that contains the class via the Workers Dashboard or via the command line.
+
+```sh
+curl -i -H "X-Auth-Email: ${EMAIL}" -H "X-Auth-Key: ${AUTH_KEY}" "https://api.cloudflare.com/client/v4/accounts/${ACCOUNT_TAG}/workers/scripts/${SCRIPT_NAME}" -X PUT -H "Content-Type: application/javascript" --data "`cat durable_object_example.js`"
```

Comment:
We might encourage people to use API tokens rather than auth keys, which just requires a header like "Authorization: Bearer <token>". Probably worth checking out what other parts of the docs use.